packages/agreement-platformstate-writer/src/consumerServiceV2.ts

```diff
@@ -1,27 +1,279 @@
+      const agreement = parseAgreement(msg.data.agreement);
+      const primaryKey = makePlatformStatesAgreementPK(agreement.id);
+      const agreementEntry = await readAgreementEntry(
+        primaryKey,
+        dynamoDBClient
+      );
+
+      const pkCatalogEntry = makePlatformStatesEServiceDescriptorPK({
+        eserviceId: agreement.eserviceId,
+        descriptorId: agreement.descriptorId,
+      });
+      const catalogEntry = await readCatalogEntry(
+        pkCatalogEntry,
+        dynamoDBClient
+      );
+      if (!catalogEntry) {
+        // TODO double-check
+        throw genericInternalError("Catalog entry not found");
+      }
+
+      const GSIPK_consumerId_eserviceId = makeGSIPKConsumerIdEServiceId({
+        consumerId: agreement.consumerId,
+        eserviceId: agreement.eserviceId,
+      });
+
+      if (agreementEntry) {
+        if (agreementEntry.version > msg.version) {
+          return Promise.resolve();
+        } else {
+          await updateAgreementStateInPlatformStatesEntry(
+            dynamoDBClient,
+            primaryKey,
+            agreementStateToItemState(agreement.state),
+            msg.version
+          );
+        }
+      } else {
+        const newAgreementEntry: PlatformStatesAgreementEntry = {
+          PK: primaryKey,
+          state: agreementStateToItemState(agreement.state),
+          version: msg.version,
+          updatedAt: new Date().toISOString(),
+          GSIPK_consumerId_eserviceId,
+          GSISK_agreementTimestamp:
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            agreement.stamps.activation!.when.toISOString(),
+          agreementDescriptorId: agreement.descriptorId,
+        };
+
+        await writeAgreementEntry(newAgreementEntry, dynamoDBClient);
+      }
+
+      const updateLatestAgreementOnTokenStates = async (
+        catalogEntry: PlatformStatesCatalogEntry
+      ): Promise<void> => {
+        if (
+          await isLatestAgreement(
+            GSIPK_consumerId_eserviceId,
+            agreement.id,
+            dynamoDBClient
+          )
+        ) {
+          // token-generation-states only if agreement is the latest
+          const GSIPK_eserviceId_descriptorId = makeGSIPKEServiceIdDescriptorId(
+            {
+              eserviceId: agreement.eserviceId,
+              descriptorId: agreement.descriptorId,
+            }
+          );
+
+          await updateAgreementStateAndDescriptorInfoOnTokenStates({
+            GSIPK_consumerId_eserviceId,
+            agreementState: agreement.state,
+            dynamoDBClient,
+            GSIPK_eserviceId_descriptorId,
+            catalogEntry,
+          });
+        }
+      };
+
+      await updateLatestAgreementOnTokenStates(catalogEntry);
+
+      const updatedCatalogEntry = await readCatalogEntry(
+        pkCatalogEntry,
+        dynamoDBClient
+      );
+      if (!updatedCatalogEntry) {
+        // TODO double-check
```

Comment:
same considerations

Reply:
https://github.com/pagopa/interop-be-monorepo/pull/987#issuecomment-2446459262